### PR TITLE
Consider TypeBox post

### PR DESCRIPTION
**Pull request by tmcw**
I was going to write this just as a post on https://macwright.com/micro/ but decided it might make good content for this blog too, so… thoughts? Should we publish this?

### CODE REVIEW

File: src/content/blog/consider-typebox.mdx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    params: Type.Object({
+      handle: Type.String(),
+      name: Type.String(),
+    }),
+  },
+  // …
+});
+```
+
+Let's look at the `querystring` type (and [here's a Val that shows doing just this](https://www.val.town/v/tmcw/typeboxExample)):
```
Comment (stevekrouse) [inline src/content/blog/consider-typebox.mdx:56]: This val looks private

File: src/content/blog/consider-typebox.mdx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  v: Type.Optional(Type.Integer({ minimum: 0 })),
+});
+
+console.log(JSON.stringify(querystring));
+// {"type":"object","properties":{"v":{"minimum":0,"type":"integer"}}}
+```
+
+TypeBox objects are JSON Schema objects with extra in-memory magic!
+
+A fun way to think about TypeScript is that it does for JSON Schema what [drizzle does for SQL](https://orm.drizzle.team/) - it's a nice, TypeScript-friendly builder syntax for an existing spec that tries to expose all of the spec's power. That's not _all_ TypeBox is - the
```
Comment (stevekrouse) [inline src/content/blog/consider-typebox.mdx:69]: Seems like the last sentence of this paragraph is unfinishe

File: src/content/blog/consider-typebox.mdx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+I am a strong believer in Carlos Bueno’s [mature optimization handbook](https://carlos.bueno.org/optimization/), which dictates that:
+
+> The trickiest part of speeding up a program is not doing it, but deciding whether it’s worth doing at all.
+
+And, later:
+
+> Before you can optimize anything you need a way to measure what you are optimizing. Otherwise you are just shooting in the dark. Before you can measure you need a clear, explicit statement of the problem you are trying to solve. Otherwise, in a very real sense, you don’t know what you are doing.
+
+So, I keep in mind that everything in the application that is not the bottleneck, is not the bottleneck. I know in my mind that there's a performance difference between different kinds of loops in JavaScript, but when I see an `Array.map` that's not strictly necessary but it's iterating over 100 items at most, it's best to take it easy.
```
Comment (stevekrouse) [inline src/content/blog/consider-typebox.mdx:142]: "take it easy" here is a bit ambiguous

**Comment by stevekrouse** [comment]
Great post! 

It feels like you take a while to get to the main point: We want to autogenerate an OpenAPI Spec, and Zod is *too* expressive. We need something that maps directly onto OpenAPI and TypeBox does that. (I wonder if a Venn Diagram would help here? Probably not, but maybe?)

But I like your meandering casual style so feel free to ignore that comment and post away :)

File: src/content/blog/consider-typebox.mdx [hunk 4]
```diff
@@ -0,0 +1,164 @@
+---
+title: Consider TypeBox
```
[the remark this answers is not in the document]
Comment (tmcw) [reply src/content/blog/consider-typebox.mdx]: Retitled to: "Zod is amazing. Here’s why we're also using TypeBox"

**Comment by tmcw** [comment]
Yeah, good point - put a lede much earlier in it to help with that.

File: src/content/blog/consider-typebox.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevekrouse) [reply src/content/blog/consider-typebox.mdx]: Or: "Zod is amazing, but not for OpenAPI"
Or: "Zod is amazing, but TypeBox is for OpenAPI"

[the remark this answers is not in the document]
Comment (stevekrouse) [reply src/content/blog/consider-typebox.mdx]: Feel free to ignore :)

**Comment by tmcw** [comment]
All right, I think this is basically ready to ship. I think we could push this live today as a low-pressure devlog post and I'll release the other post tomorrow with screenshots / a video after we've shipped Townie UI improvements @stevekrouse ?

**Pull request merged**